docs/decisions/broker-trigger.md

```diff
@@ -0,0 +1,137 @@
+# Broker and Trigger model (for 0.5.0)
+
+Decision date: 6 March 2019
```

Comment:
Let's a a list of terms/components with a one-liner for each one - e.g.:
- Event Producer - creates events based on occurrences. E.g. github
- Event Consumer - destination for events from an Event Producer. E.g. the application processing an event. Typically, this is the final "sink" in the processing flow of an Event.
- Event Source - a helper entity that can be used to connect an Event Producer to the Kn eventing system and a destination (sink) for those events (e.g. an Event Consumer or a Broker)
- Event Source Service - a Knative Service that is the endpoint to which Event Producers send events. This is created by the Event Source
- Sink - a generic term to mean the destination for an event being sent from a component.
- Broker - an entity that can be a Sink. Subscriptions can be created for the Broker to register interest in Events that flow through the Broker. Brokers are typically used to provide loose coupling between the incoming events and Event Consumers - e.g. it may buffer/queue events.
- Trigger/Subscription - a registered interest in events that flow through a Broker. This will typically include a Sink (where to send the Events) and an optional filter to control which Events are sent. It may buffer/queue events.

Not sure about both the Broker and Trigger buffering things.... did we talk about that? I noticed below it talks about the Trigger doing the buffering.

Reply:
I think including this would help. In particular I like is that the terms have a concrete example associated with them.

I don't think 'Event Source Service' is needed. Or if it is included, it should not mandate that it is a Knative `Service` (e.g. it could be a K8s `Deployment`).

Event Producer should also specify that the event produced does not need to be a CloudEvent.

Reply:
Conceptually, each Trigger has an independent buffer. The implementation of where the buffer lives in the data plane is a bit mushier, as I expect that the Broker and Trigger will collapse into the same data-plane objects in many cases.

Reply:
I've added some updated definitions (for example, the "Event Source Service" is called a "Receive Adapter", and might be a Deployment or StatefulSet instead of a Knative Service, depending on the Event Source implementation).

I left out the following two definitions, because they are defined later in the document in more detail:

- **Broker**: an event dispatch object which acts as an event Sink. Triggers can
  be created on the Broker to register interest in Events that are delivered to
  the Broker. Brokers are used to provide loose coupling and content-based
  routing between incoming events and Event Consumers - e.g. it may
  buffer/queue events.
- **Trigger** a registered interest in events that flow through a Broker. This
  will typically include a Sink (where to send the Events) and an optional
  filter to control which Events are sent. It may buffer/queue events.